src/stan_inference.jl

```diff
@@ -49,11 +48,16 @@ end
 function stan_inference(prob::DiffEqBase.DEProblem,t,data,priors = nothing;alg=:rk45,
                             num_samples=1000, num_warmup=1000, reltol=1e-3,
                             abstol=1e-6, maxiter=Int(1e5),likelihood=Normal,
-                            vars=(StanODEData(),InverseGamma(3,3)),nchains=1)
+                            vars=(StanODEData(),InverseGamma(3,3)),nchains=1,
+                            obsvbls = 1:size(data, 1), sample_u0 = false, diffeq_string = "")
```

Comment:
what's the application for having `diffeq_string` here? Just curious. Might be easier to have a default of `nothing`.

Reply:
I will make `diffeq_string` default to `nothing`. This is my use case.

```
diffeq_string = """
real af(real x, real theta, real slope) {
    return 1 / (1 + exp(-(x - theta) / slope));
}
real[] sho(real t, real[] u, real[] p, real[] x_r, int[] x_i) {
    real du[12];
    real x1 = u[1]; real y1 = u[2]; real x2 = u[3]; real y2 = u[4]; real m1 = u[5]; real k1 = u[6];
    real x3 = u[7]; real y3 = u[8]; real x4 = u[9]; real y4 = u[10]; real  m2 = u[11]; real k2 = u[12];
    real r = p[1]; real s1 = p[2]; real s2 = p[3]; real eps = p[4]; real eta = p[5]; real w_12 = p[6];
    real w_21 = p[7]; real theta = p[8]; real v_syn = p[9]; real wk1 = p[10]; real g1 = p[11]; real g2 = p[12];
    real tau = 0.05; real taum = 0.2; real tau2 = 0.5; real tauk = 0.5;
    real wk2 = 1; real w_34 = 0; real w_43 = 0; real eta2 = 6;
    real W = p[12 + bin_search(ceil((t - 1) / $dt), 0, 1000)];
    du[1] = (r + W - s2 * x1^3 + s1 * x1 - eps * y1 - w_12 * af(x2, theta, 0.1) * (x1 - v_syn) + g1 * (x3 - x1) - eta * k1) / tau;
    du[2] = -y1 + af(x1, 0, 0.5);
    du[3] = (r + W - s2 * x2^3 + s1 * x2 - eps * y2 - w_21 * af(x1, theta, 0.1) * (x2 - v_syn) + g2 * (x4 - x2) + eta * k1) / tau;
    du[4] = -y2 + af(x2, 0, 0.5);
    du[5] = (-m1 + wk1 * (af(x1, theta, 0.1) - af(x2, theta, 0.1))) / taum;
    du[6] = (-k1 + m1) / tauk;
    du[7] = (x3 - x3^3 - w_34 * af(x4, theta, 0.1) * (x3 - v_syn) + g1 * (x1 - x3) - eta2 * k2) / tau2;
    du[8] = -y3 + af(x3, 0, 0.5);
    du[9] = (x4 - x4^3 - w_43 * af(x3, theta, 0.1) * (x4 - v_syn) + g2 * (x2 - x4) + eta2 * k2) / tau2;
    du[10] = -y4 + af(x4, 0, 0.5);
    du[11] = (-m2 + wk2 * (af(x3, theta, 0.1) - af(x4, theta, 0.1))) / taum;
    du[12] = (-k2 + m2) / tauk;
    return du;
}
"""
res_stan = stan_inference(
    prob, (0:dt:tend) .+ 1, data, priors,
    num_samples = 5000, num_warmup = 500,
    vars =(DiffEqBayes.StanODEData(), InverseGamma(3, 2)), 
    obsvbls = 6:6, sample_u0 = true, diffeq_string = diffeq_string
)
```

`real W = p[12 + bin_search(ceil((t - 1) / $dt), 0, 1000)];` cannot be handled easily in Julia.

Reply:
yeah, interesting. The new ModelingToolkit form will handle the embded function, but it won't call the Stan `bin_search` function.